### PR TITLE
[primTorch] Adds random operations

### PR DESCRIPTION
**Pull request by mruberry**
This PR...

**Issues Found**
- https://github.com/pytorch/pytorch/issues/78058
- https://github.com/pytorch/pytorch/issues/78054
- https://github.com/pytorch/pytorch/issues/78053
- https://github.com/pytorch/pytorch/issues/78050
- https://github.com/pytorch/pytorch/issues/77932

**Testing**
- disables stride consistency checks in test_ops and test_meta pending resolution of https://github.com/pytorch/pytorch/issues/78050
- skips chalf in reference tests (addressing https://github.com/pytorch/pytorch/issues/78054)
- splits test test_python_reference_consistency in one test for the ctx where torch.foo is torch.foo, and another for when torch.foo is refs.foo
- updates test names to be more natural and consistent:
  - test_python_reference_errors -> test_python_ref_errors
  - test_python_reference_consistency -> test_python_ref and test_python_ref_torch_fallback 
  - test_python_reference_meta_functions -> test_python_ref_meta
  - test_reference_testing -> test_numpy_ref
- updates test_python_ref and test_python_ref_torch_fallback to check that the reference is more accurate than the torch op if the reference and torch op results are not close, a warning is raised when this occurs (addressing https://github.com/pytorch/pytorch/issues/77687)
- adds reference inputs for broadcast_tensors
- Updates the "fill_" OpInfo to "fill", adding a NumPy reference and making it an elementwise unary operator
- Adds 1D no element sample inputs to the cat OpInfo and updates the NumPy reference to handle them and type promotion correctly
- Adds reference inputs for elementwise ternary operations, like clamp
- Adds a NumPy reference for clamp
- Adds reference inputs to where's OpInfo
- Makes softplus an elementwise unary OpInfo
- Removes the great majority of Python reference OpInfo skips and xfails due to the above test changes
- Adds Python reference OpInfos for fill, dropout, clamp, broadcast_tensors, and where

**Prims**
- adds the fill, empty_strided, and uniform prims
- removes the empty, empty_like, full, and full_like prims -- these are now references that use empty_strided and fill
- renames the "concatenate" and "select" prims to "cat" and "where", respectively, to be consistent with PyTorch
- extends the `_elementwise_meta` operation to accepts tensors that don't participate in type promotion, like the `cond` tensor in `where`
- fixes a bug in the stride propagation of broadcast_in_dim
- moves some error checks from prims.cat to prims.where to refs.cat and refs.where, respectively, consistent with our new policy of doing as much error checking in the ref as possible

**Utils**
- adds the canoicalize_device, extract_shape, and extract_shape_from_varargs helpers
- adds the elementwise_unary_scalar_wrapper -- this allows elementwise unary operators to take and return scalar values (ex. refs.sin(1) will return .84...)

**Refs**
- adds the fill, broadcast_tensors, clamp, empty_strided, ones, zeros, and uniform references
- adds the nn.functional.dropout reference
- fixes refs.cat to handle 1D tensors with no inputs consistent with eager mode

### CODE REVIEW

File: test/test_ops.py [hunk 1]
```diff
@@ -366,47 +370,119 @@ def _to_tensormeta(x):
...
+            for a, b in zip(tree_flatten(ref_result)[0], tree_flatten(precise_result)[0]):
+                ref_distance = ref_distance + _distance(a, b)
+
+            torch_distance = 0
+            for a, b in zip(tree_flatten(torch_result)[0], tree_flatten(precise_result)[0]):
+                torch_distance = torch_distance + _distance(a, b)
+
+            msg = f"Reference result was farther ({ref_distance}) from the precise \
+                    computation than the torch result was ({torch_distance})!"
+            self.assertTrue(ref_distance <= torch_distance, msg=msg)
```
Comment (ngimel) [inline test/test_ops.py:456]: I don't think ref_distance is always weakly less than torch_distance, it's ok to be larger with some tolerance?

File: test/test_ops.py [hunk 2]
```diff
@@ -366,47 +370,119 @@ def _to_tensormeta(x):
...
+
+        # Reports numerical accuracy discrepancies
+        if ex is not None:
+            msg = "Test passed because the reference was more accurate than the torch operator."
+            print(msg)
```
Comment (ngimel) [inline test/test_ops.py:461]: should this be a warning? Pytest hides stdout of passing tests, so it will be hard to see for people using pytest

File: torch/_prims/__init__.py [hunk 3]
```diff
@@ -253,16 +257,21 @@ def _elementwise_meta(
...
+    utils.check_same_device(*args_, allow_cpu_scalar_tensors=True)
+    utils.check_same_shape(*args_, allow_cpu_scalar_tensors=True)
+
+    strides = utils.compute_elementwise_output_strides(*args_)
 
     tensor = None
     scalar_tensor = None
     number = None
-    for arg in args:
+    for arg in args_:
```
Comment (ngimel) [inline torch/_prims/__init__.py:274]: this would potentially set `tensor` to something with wrong dtype (from `args_with_different_types`)

File: torch/_prims/__init__.py [hunk 4]
```diff
@@ -242,7 +244,9 @@ class ELEMENTWISE_PRIM_TYPE_PROMOTION_KIND(Enum):
 
 # TODO: implement dtype validation here, too, or on the corresponding refs
 def _elementwise_meta(
-    *args, type_promotion: ELEMENTWISE_PRIM_TYPE_PROMOTION_KIND
+    *args,
+    type_promotion: ELEMENTWISE_PRIM_TYPE_PROMOTION_KIND,
+    args_with_different_dtypes: Tuple[TensorLikeType, ...] = None,
```
Comment (ngimel) [inline torch/_prims/__init__.py:249]: `args_with_fixed_types`?

File: torch/_prims/__init__.py [hunk 5]
```diff
@@ -1589,35 +1613,35 @@ def _rev_meta(a: TensorLikeType, dims: DimsSequenceType) -> TensorLikeType:
...
-        a, b, type_promotion=ELEMENTWISE_PRIM_TYPE_PROMOTION_KIND.DEFAULT
+        a,
+        b,
+        type_promotion=ELEMENTWISE_PRIM_TYPE_PROMOTION_KIND.DEFAULT,
+        args_with_different_dtypes=(pred,),
     )
 
 
-def _select_aten(pred: Tensor, a: Tensor, b: Tensor) -> Tensor:
+def _where_aten(pred: Tensor, a: Tensor, b: Tensor) -> Tensor:
```
Comment (ngimel) [inline torch/_prims/__init__.py:1628]: out of curiosity, why do we need this helper, and not just use `torch.where` in make_prim?

File: torch/_prims/__init__.py [hunk 6]
```diff
@@ -1927,37 +1953,42 @@ def _empty_aten(
...
+    dtype: torch.dtype,
+    device: torch.device,
+    requires_grad: bool,
+) -> TensorLikeType:
+    return TensorMeta(shape=shape, strides=strides, dtype=dtype, device=device)
 
 
-def _empty_like_aten(
-    a: Tensor, *, dtype: torch.dtype, device: torch.device, requires_grad: bool
+def _empty_strided_aten(
```
Comment (ngimel) [inline torch/_prims/__init__.py:1968]: same question here?

File: test/test_ops.py (hunk 1 above)
Comment (mruberry) [reply test/test_ops.py:456]: I was also thinking about this but wasn't happy with any immediate ideas I had -- cool if I add a TODO comment?

File: test/test_ops.py (hunk 2 above)
Comment (mruberry) [reply test/test_ops.py:461]: Sure -- warning it is

File: torch/_prims/__init__.py (hunk 4 above)
Comment (mruberry) [reply torch/_prims/__init__.py:249]: Yeah that's way better -- fixed

File: torch/_prims/__init__.py (hunk 3 above)
Comment (mruberry) [reply torch/_prims/__init__.py:274]: Great catch - fixed

File: torch/_prims/__init__.py (hunk 5 above)
Comment (mruberry) [reply torch/_prims/__init__.py:1628]: Great point -- I was just on automatic mode -- fixed!

File: torch/_prims/__init__.py (hunk 6 above)
Comment (mruberry) [reply torch/_prims/__init__.py:1968]: Yep -- fixed

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -3794,6 +3813,22 @@ def sample_inputs_fill_(op_info, device, dtype, requires_grad, **kwargs):
...
+        t.fill_(value)
+    return t
+
+def _fill_sample_kwargs(device, dtype, input):
+    if dtype is torch.bool:
+        value = True
+    else:
+        value = 3
+
+    return ({'value': value}, {'value': value})
```
Comment (ngimel) [inline torch/testing/_internal/common_methods_invocations.py:3831]: why 2 tuple elements?

**Review by ngimel — APPROVED** [review]
Great

File: torch/testing/_internal/common_methods_invocations.py (hunk 7 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:3831]: This is super weird and we could sugar over this, but it's because we sometimes pass different arguments to the NumPy op, so we have "torch kwargs" and "NumPy kwargs" here

**Comment by suo** [comment]
Actually this looks like the proper classification is "nosignal"--only slow tests broke.

File: torch/_refs/special/__init__.py [hunk 8]
```diff
@@ -33,10 +35,10 @@
...
-def logit(self: Tensor, eps: Optional[float] = None) -> Tensor:
+def logit(self: TensorLikeType, eps: Optional[float] = None) -> TensorLikeType:
     if eps is None:
         eps = -1.0
     lo = eps
     hi = 1 - eps
-    self = torch.clamp(self, lo, hi)
-    return (self / (1 - self)).log()
+    self = refs.clamp(self, lo, hi)
+    return refs.log(refs.true_divide(self, refs.sub(1, self)))
```
Comment (ezyang) [inline torch/_refs/special/__init__.py:44]: @mruberry Given that the context manager exists now, we should *prefer* using the `torch` API calls as this ensures that the decomposition in question is using the limited API supported by torch and not the expanded API from refs. Is this just to work around the local problem that ref consistency tests don't work? I'd much rather we dupe the tests in that case.

Comment (mruberry) [reply torch/_refs/special/__init__.py:44]: It's that the meta tests weren't working; I did duplicate the consistency tests
